Fixed dropwizard jersey duplicate path logging

afrin216 opened the pull request:
So this was a tiny bug I discovered while using dropwizard (which is awesome by the way) on our project. The problem being if there are two separate functions present exposing same path in jersey but supporting different media type, then dropwizard logs duplicate paths.
Eg:
```
@Get
@Path("/abc")
@Produces(MediaType.TEXT_HTML)
public Respone functionA(){
   //some code produces html output
}

@Get
@Path("/abc")
@Produces(MediaType.APPLICATION_JSON)
public Respone functionB(){
   //some code produces json output
}
```
Now dropwizard logs the following
```
GET /abc MyResourceClass
GET /abc MyResourceClass
```
This PR fixes that and logs just one line by using a TreeSet instead of List. Its passive and doesn't affect anything outside this class (There could be an enhancement to log the one line and list all supported content types, for which I might do a PR later).

Thanks and continue being awesome guys!

nickbabcock commented on dropwizard-jersey/src/test/java/io/dropwizard/jersey/DropwizardResourceConfigTest.java:
```diff
@@ -118,6 +120,22 @@ public void logsNestedEndpoints() {
+                + "    GET     /callme (io.dropwizard.jersey.DropwizardResourceConfigTest.TestDuplicateResource)%n");
+
+        assertThat(rc.getEndpointsInfo()).contains(expectedLog);
+        assertThat(rc.getEndpointsInfo()).doesNotContain(shouldNotContainLog);
```
Would it make sense to have

```java
assertThat(rc.getEndpointsInfo())
        .containsOnlyOnce()("    GET     /callme (io.dropwizard.jersey.DropwizardResourceConfigTest.TestDuplicateResource"))
```

nickbabcock commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/DropwizardResourceConfig.java:
```diff
@@ -213,18 +210,52 @@ public EndpointLogLine(String httpMethod, String basePath, Class<?> klass) {
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
```
Instead of hand writing `hashCode` can you use `Objects.hash`?

nickbabcock commented on dropwizard-jersey/src/main/java/io/dropwizard/jersey/DropwizardResourceConfig.java:
```diff
@@ -114,7 +111,7 @@ public void setUrlPattern(String urlPattern) {
     public String getEndpointsInfo() {
         final StringBuilder msg = new StringBuilder(1024);
-        final List<EndpointLogLine> endpointLogLines = Lists.newArrayList();
+        final Set<EndpointLogLine> endpointLogLines = Sets.newTreeSet();
```
`newTreeSet` accepts a `Comparator`, any reason why you couldn't reuse `EndpointComparator`?

afrin216 replied:
My bad. Reverted!

nickbabcock commented:
Hey, thanks for the PR, I like the direction (I too think it'd be neat if supported content-types were logged with the endpoints, yet that could make for some long log lines), but I had a few questions in my code comments that should be addressed.

I do wonder if anyone depends on the current behavior to know that multiple content types were registered...hmm

afrin216 commented:
@nickbabcock  Addressed your comments. Thanks.

Yeah it could make the logs a little longer. But we could organise it so that only for those endpoints with restricted return types, further info such as content-types gets printed. Eg :-
````
GET /abc MyResourceClass
        Accepts  - application/json,  text/html
        Produces - application/xml
GET /abc2 MyResourceClass
GET /abc3 MyResourceClass
GET /abc4 MyResourceClass
````

Would be useful atleast during debugging.

nickbabcock commented:
Ohhh, I personally really like that output, but I'll see if others have opinions on that.

Also you can simplify the code even more. There is no need to override `hashCode` or `equals` :smile:

afrin216 commented:
Eh right! Changes do look cleaner now. Thanks! :smile:

jplock commented:
My recommendation would be to keep this PR focused on fixing the path logging issue and if we want to change around the output format, do that in another PR targeted for a future release.

nickbabcock commented:
@jplock, absolutely -- keep pull requests focused!

Squashed and merged PR as part of c91851b

Thanks, @afrin216
